### PR TITLE
Fix opacity reset

### LINKED ISSUE #7176: Opacity settings keep resetting to high values
### Actions before raising this issue

- [X] I searched the existing issues and did not find anything similar.
- [X] I read/searched [the docs](https://github.com/cvat-ai/cvat/tree/master#documentation)

### Steps to Reproduce

1. Go to any task of the project;
2. Change Opacity settings;
2. Change frame/Make any update;
3. Opacity settings will reset to the same high values;

### Expected Behavior

Expected the settings to remain the changed values, however, they keep being reset.

### Possible Solution

We are certain this is from the server-side, because the entirety of our project was affected. No matter which video is selected, the opacity values default to high once a page update occurs. Restarting the computer, refreshing the page, saving annotations with the correct Opacity did not help.

### Context

This issue affected me by preventing accurate labelling of sensitive objects. The high opacity becomes an issue to to the inability to notice the finer borders.

### Environment

_No response_

### PR DESCRIPTION
**Pull request by klakhov**
### Motivation and context
Resolved #7176

Opacity keeps resetting after changing frame (even without the masks)
opacity-reset

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- ~~[ ] I have updated the documentation accordingly~~
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-ui/src/reducers/settings-reducer.ts [hunk 1]
```diff
@@ -12,13 +12,13 @@ import { AnnotationActionTypes } from 'actions/annotation-actions';
...
-import { ObjectState, ShapeType, DimensionType } from 'cvat-core-wrapper';
+import { clampOpacity } from 'utils/clamp-opacity';
 
 const defaultState: SettingsState = {
     shapes: {
         colorBy: ColorBy.LABEL,
-        opacity: 3,
-        selectedOpacity: 30,
+        opacity: 30,
+        selectedOpacity: 60,
```
Comment (bsekachev) [inline cvat-ui/src/reducers/settings-reducer.ts:21]: I do not think we need to update these default values

File: cvat-ui/src/utils/clamp-opacity.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { SettingsState } from 'reducers';
+
+export function clampOpacity(
+    states: ObjectState[],
+    shapes: SettingsState['shapes'],
+    job?: Job,
+): [number, number] {
+    const DEFAULT_OPACITY = 30;
+    const DEFAULT_SELECTED_OPACITY = 60;
+
```
Comment (bsekachev) [inline cvat-ui/src/utils/clamp-opacity.ts:17]:
```suggestion
    const ENHANCED_DEFAULT_OPACITY = 30;
    const ENHANCED_DEFAULT_SELECTED_OPACITY = 60;

```

**Comment by bsekachev** [comment]
Changelog?

File: changelog.d/20231127_123423_klakhov_fix_opacity_reset.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+### Fixed
+
+- Resetting appearance opacity on each `change:frame`
```
Comment (bsekachev) [inline changelog.d/20231127_123423_klakhov_fix_opacity_reset.md:3]:
```suggestion
- Shape settings **opacity** and **selected opacity** reset on each frame change
```

**Review by bsekachev — APPROVED** [review]

**Comment by bsekachev** [comment]
Now there is one more issue, when I navigate from a frame without masks to frame with masks, opacity keeps on low level and masks are not visible

image

File: cvat-ui/src/utils/clamp-opacity.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+): [number, number] {
+    const ENHANCED_DEFAULT_OPACITY = 30;
+    const ENHANCED_DEFAULT_SELECTED_OPACITY = 60;
+
+    const withMasks = states
+        .some((_state: ObjectState): boolean => _state.shapeType === ShapeType.MASK);
+    const opacity = withMasks || job?.dimension === DimensionType.DIMENSION_3D ?
+        Math.max(shapes.opacity, ENHANCED_DEFAULT_OPACITY) : shapes.opacity;
+    const selectedOpacity = withMasks || job?.dimension === DimensionType.DIMENSION_3D ?
+        Math.max(shapes.selectedOpacity, ENHANCED_DEFAULT_SELECTED_OPACITY) : shapes.selectedOpacity;
```
Comment (bsekachev) [inline cvat-ui/src/utils/clamp-opacity.ts:23]: As an optimization, I would suggest first to check if current opacity/selected opacity already more or equal than enhanced level, just skip the algorithm. Also checking job?.dimension is quicker, so, it is a thing we should probably do first

**Review by bsekachev — APPROVED** [review]

**Comment by nazmulhasan202** [comment]
I am still facing the issue raised in the link: https://github.com/cvat-ai/cvat/issues/7176 
I would really appreciate it if you could resolve it at your earliest convenience. Thanks!

**Comment by nazmulhasan202** [comment]
Today (July 22, 2024), I found out that the Opacity Resetting issue has been resolved. Thanks a lot to the CVAT team.

**Pull request merged**